exploration/pattern-exterior-whitespace.md

```diff
@@ -0,0 +1,200 @@
+# Pattern Exterior Whitespace
+
+Status: **Accepted**
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@aphillips</dd>
+		<dd>@eemeli</dd>
+		<dt>First proposed</dt>
+		<dd>2023-10-02</dd>
+		<dt>Pull Request</dt>
+		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/487">#487</a></dd>
+	</dl>
+</details>
+
+## Objective
+
+_What is this proposal trying to achieve?_
+
+The WG is discussing how to handle "pattern exterior" whitespace,
+which is ASCII whitespace (tab, CR, LF, or U+0020) that is **_part_** of the pattern
```

Comment:
This definition of "ASCII whitespace" differs from that of the [web platform](https://infra.spec.whatwg.org/#ascii-whitespace) (which additionally includes U+000C FORM FEED) and of [Unicode](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%5B%3AASCII%3A%5D%26%5B%3AWhite_Space%3DYes%3A%5D%5D&g=&i=) (which additionally includes both U+000C FORM FEED and U+000B VERTICAL TABULATION).

Reply:
We omitted form feed and vtab, which are ASCII whitespace characters. We are consistent with JSON, HTML, and CSS's definition of whitespace. Many other languages also observe this definition (Java and I believe C++, although not sure about the latter) Is there a technical reason for us to permit form feed and vtab into our whitespace definition? If we don't permit them in our whitespace production, they would not be "pattern exterior" (for many of the cases in this document). We don't currently allow them in expressions, declarations and the like.

Reply:
> We omitted form feed and vtab, which are ASCII whitespace characters. We are consistent with JSON, HTML, and CSS's definition of whitespace.

This is consistent with JSON ([ref](https://www.rfc-editor.org/rfc/rfc8259#page-6)), but not with HTML or CSS by my reading of those specifications:
* HTML uses the Infra Standard definition of [ASCII whitespace](https://infra.spec.whatwg.org/#ascii-whitespace) (U+0009 TAB, U+000A LF, **U+000C FF**, U+000D CR, or U+0020 SPACE) that I linked to above ([example](https://html.spec.whatwg.org/multipage/dom.html#content-models)).
* CSS is more complicated because [preprocessing](https://drafts.csswg.org/css-syntax/#input-preprocessing) replaces U+000D CARRIAGE RETURN and **U+000C FORM FEED** with U+000A LINE FEED, but basically boils down to the union of those with U+0009 CHARACTER TABULATION and U+0020 SPACE, aligning with HTML ([ref](https://drafts.csswg.org/css-syntax/#newline)).

> Many other languages also observe this definition (Java and I believe C++, although not sure about the latter) Is there a technical reason for us to permit form feed and vtab into our whitespace definition? If we don't permit them in our whitespace production, they would not be "pattern exterior" (for many of the cases in this document). We don't currently allow them in expressions, declarations and the like.

I'm not arguing for inclusion or exclusion of the extra control characters, just for clarity. "ASCII whitespace" is a term of art in the web platform that includes U+000C FORM FEED, and in Unicode implies an interpretation like "code points that are both ASCII and White_Space" that includes U+000C FORM FEED and U+000B VERTICAL TABULATION. This text was therefore misleading, but https://github.com/unicode-org/message-format-wg/pull/487/commits/cd188aa7a8aa4d936b14dd2fb3310fd5072e1944 resolved that. :+1:

Reply:
Can I get a "ship-it" (approval)?